### PR TITLE
bpo-25658: Implement PEP 539 for Thread Specific Storage (TSS) API

### PR DESCRIPTION
**Pull request by ma8ma**
~~Do not merge until approved the PEP final!~~
Update (2017-09-09): PEP 539 has been accepted.

## Thread Specific Storage (TSS) API

Implement [PEP 539][] for Thread Specific Stroage (TSS) API: it is a new Thread Local Storage (TLS) API to CPython which would supersede use of the existing TLS API within the CPython interpreter, while deprecating the existing API.

### Highlights of changes

- Add Thread Specific Storage (TSS) API
- Add test for transitions of thread key state
- Mark deprecation to Thread Local Storage (TLS) API
- Replace codes that used TLS API with TSS API
- Add the TSS API in the Windows stable ABI (https://github.com/python/cpython/pull/1362#issuecomment-299707811)
- ~~[TEMP] Add an example function to ``xxlimited`` module~~

[PEP 539]: https://www.python.org/dev/peps/pep-0539/

https://bugs.python.org/issue25658

### CODE REVIEW

**Comment by AraHaan** [comment]
Uh are you rebasing? I made that mistake before. The latest dev guide says to not rebase but include all commits which would then be merged as a single commit.

**Comment by ma8ma** [comment]
@AraHaan Yes, PR is single commit. I confirmed AppVeyor build failed, and retried push in a minute. But it has failed again, I need help for Windows platform...

**Comment by AraHaan** [comment]
It seems to be something with the solution or project settings or even AppVeyor's setting. Will investigate more when I can use the internet on my Inspiron 1545 that has Windows 7 Ultimate SP1 x64 and VS2015 installed. It seems that in the project settings is set to use the Windows sdk for 8.1 instead of for 10. Also when targeting for 10.0 I recommend staticallylinking the ucrt for the versions of Windows that can't install it (Vista). Also it allows flexibility and usability on fresh installs of 7 and Vista do that is another reason why I recommend it.

File: Include/pythread.h
Comment (ncoghlan) [inline Include/pythread.h]: This comment should explain that the type layout is public (despite being nominally opaque) to allow for static declarations in API clients.

It should be made fully opaque when `Py_LIMITED_API` is defined, though: https://docs.python.org/3/c-api/stable.html

File: Include/pythread.h
Comment (ncoghlan) [inline Include/pythread.h]: To allow for use in the `Py_LIMITED_API` case, the public APIs that currently accept `Py_tss_t` by value will need to accept it as a pointer instead.

The alternative would be to make the `Py_tss_t` struct layout part of the stable ABI, but I'm reluctant to do that due to the `NATIVE_TLS_KEY_T` field: it would keep us from switching from `int` to a different native type for platforms that don't currently have a native implementation, but gain one in the future.

Following that chain of thought further, I guess fully supporting Py_LIMITED_API would require an extra pair of memory management functions (`PyThread_tss_alloc` and `PyThread_tss_free`), as with the struct definition hidden, API clients won't be able to use static variables.

File: Python/thread.c
Comment (ncoghlan) [inline Python/thread.c]: For the second sentence here, I'd suggest "While the native thread-local storage implementations have been retained, this fallback implementation is now just a compatibility wrapper around the TSS API."

File: Python/thread_foobar.h
Comment (ncoghlan) [inline Python/thread_foobar.h]: Does PEP 7 express  a preference regarding empty function definitions? If it doesn't we should at least make sure the TLS/TSS code is self-consistent (one of the later changes has the opening and closing brace on the same line).

**Review by ncoghlan — CHANGES_REQUESTED** [review]
This looks excellent to me as a solution for the standard ABI, but I realised in reviewing it that the stable ABI poses a potential problem: we don't want to expose the `Py_tss_t` definition in that case, but making the API usable with an opaque type definition would require a few changes (details inline).

File: Include/pythread.h
Comment (embray) [reply Include/pythread.h]: I'm not super-familiar with the constraints regarding `Py_LIMITED_API`.  Addressing this will require an update to the PEP as well.

I wouldn't mind if the functions in question were changed to take a pointer argument instead of passing the struct by value just for consistency's sake if nothing else.

File: Python/thread_foobar.h
Comment (embray) [reply Python/thread_foobar.h]: Not that I can tell, in which case the most consistent style would be that demonstrated here.

File: Include/pythread.h
Comment (ncoghlan) [inline Include/pythread.h]: For consistency, this should be `NATIVE_TSS_KEY_T`

File: Include/pythread.h
Comment (ma8ma) [reply Include/pythread.h]: Okay, I start work to modify to full ``Py_LIMITED_API`` on the weekend.

File: Python/thread_foobar.h
Comment (ma8ma) [reply Python/thread_foobar.h]: Sure, I'm going to make formatting even. By the way, would I create an issue for removing ``Python/thread_foobar.h``? Although I'm not sure history of thread_foobar.h, it seems to me that is dead code at present.

Comment (ncoghlan) [reply Python/thread_foobar.h]: @ma8ma Aye, a separate issue and python-dev thread for that sounds sensible to me. The question to the mailing list can just be something simple like "This looks like dead code and we want to delete it, does anyone have a reason we should keep it?"

Comment (embray) [reply Python/thread_foobar.h]: I had assumed it was a dummy implementation for testing / demonstration purposes.  It might be an easier way to understand the requirements of the threading API without being bogged down with the details of specific implementations.  That said, I don't know how useful that is in practice. I don't think I ever looked at it, opting instead to look at the pthread implementation since pthreads are already familiar...

File: Include/pythread.h
Comment (embray) [reply Include/pythread.h]: With a `PyThread_tss_alloc`, would the need for `Py_tss_NEEDS_INIT` to be public go away?  After all, if one used `PyThread_tss_alloc` to allocate memory for a struct, what else would they even initialize it with?  It would remove an extra step to just do that as part of `PyThread_tss_alloc` and remove that implementation detail.  

It should just be clear that just because a TSS key has been *allocated* doesn't mean it's been *created*.  Also, `PyThread_tss_is_created` is still available to test whether or not a key has been created.

File: Python/thread_foobar.h
Comment (ma8ma) [reply Python/thread_foobar.h]: Well, in fact, I think it makes the loading to maintainers. I posted to python-dev.

File: Include/pythread.h
Comment (ncoghlan) [reply Include/pythread.h]: static storage allocation would still be supported for the normal (non-limited API) case, so the public `Py_tss_NEEDS_INIT` initializer would still be needed.

**Comment by ma8ma** [comment]
In Windows, the module-definition (.def) file for limited API (python3.dll) doesn't have the TSS API (also the TLS API), therefore, ``xxlimited`` module build occurs [link error][appveyor]. ~~Would I append TSS API to [python3.def][python3_def] to update stable ABI?~~

**Edit (2017-05-13):** Add WIP commit for Windows ABI
**Edit (2017-08-31):** ~~As far as someone doesn't show a necessary reason, I'd revert WIP commit that adding TSS API into Windows stable ABI. (i.e. back to same support level as the existing TLS API)~~
**Update (2017-09-02):** ~~Revert a temporary commit which supports TSS API in Windows stable ABI, because the existing TLS API is not provided in Windows stable ABI. TSS API support goes to maintain the status quo.~~
**Update (2017-09-18):** Add the commit again (see https://github.com/python/cpython/pull/1362#issuecomment-330107896).

[PCbuild description](https://github.com/python/cpython/tree/master/PCbuild)

[appveyor]: https://ci.appveyor.com/project/python/cpython/build/3.7.0a0.1850
[python3_def]: https://github.com/python/cpython/blob/master/PC/python3.def

**Review by eamanu — APPROVED** [review]
All changes agrees with PEP7

File: Include/pythread.h
Comment (zooba) [inline Include/pythread.h]: Please don't use designated initializers in header files - they prevent people using Python in C++.

File: Include/pythread.h
Comment (zooba) [inline Include/pythread.h]: Why is this necessary? Just use an `int` for the initialized flag and avoid the extra dependency - every extra header can be a big deal for non-desktop compilers.

File: Include/pythread.h
Comment (zooba) [inline Include/pythread.h]: This should do a version check on the macro, since these APIs are not in the limited API prior to 3.7.

**Review by zooba — COMMENTED** [review]
I'd very much like to see the header files cleaned up to reduce potential for compatibility issues, but otherwise the implementation looks fine to me.

File: Include/pythread.h
Comment (ma8ma) [reply Include/pythread.h]: Yeah, I'll do so.

File: Include/pythread.h
Comment (ma8ma) [reply Include/pythread.h]: That's true, I'd enclose TSS API declaration.

File: Include/pythread.h
Comment (ma8ma) [reply Include/pythread.h]: Aye, I'd change PyThread_tss_is_created declaration. First of all, I report your proposal to python-dev and python/peps#382.

Comment (embray) [reply Include/pythread.h]: I prefer `bool` myself, but Steve has a point here. There's not much advantage if it means bringing in a header that wouldn't otherwise be necessary, and isn't (currently) used anywhere else in Python's headers.  If we change this it will require a minor update to the PEP.

File: Include/pythread.h
Comment (embray) [reply Include/pythread.h]: I'm not exactly sure what the alternative would be...?

Comment (ncoghlan) [reply Include/pythread.h]: Using a partial positional initializer: `#define Py_tss_NEEDS_INIT   {false,}`

The struct is simple enough that we can live without the extra clarity of repeating the field name for the benefit of better C++ compatibility.

**Comment by ma8ma** [comment]
As above, PEP 539 has been accepted, ~~but [WIP] has been labeled yet~~.
By the way, I wrote a commit message that summarizes this PR, because this PR has too much commits. The changes won't be enumerated one by one.
```
bpo-25658: Implement PEP 539 for Thread Specific Storage (TSS) API

Add a new Thread Local Storage (TLS) API to CPython which would
supersede use of the existing TLS API within the CPython interpreter,
while deprecating the existing API.  The change is motivated
primarily by the fact that the old API uses int to represent TLS keys
across all platforms, which is neither POSIX-compliant, nor portable
in any practical sense.  This change covers not only the new API
implementation features, but also the client code updates needed to
replace the existing TLS API with the new TSS API.
```

File: Python/pystate.c
Comment (ma8ma) [inline Python/pystate.c:54]: I have another idea instead of assignment by instant variable; provide a macro or inline function which ready the key (put into `Include/pythread.h`?). Edit: The name `_READY` is different behavior to `PyUnicode_READY` that returns a status code. `Py_tss_INITIALIZE (_PyThread_tss_initialize)` is more make sense?
```c
#ifndef Py_LIMITED_API
/* You *must not* use except for the case of initializing the key
   in the startup sequence! */
#define Py_tss_READY(key) do { (key)._is_initialized = 0; } while(0)
static inline void
PyThread_tss_ready(Py_tss_t *key) {
    key->_is_initialized = 0; 
}
#endif
```
In addition, calling `PyThread_tss_delete` gives to reset the key's initialization state, but it will shake the create/use/delete cycle. I don't like it.

Comment (ncoghlan) [reply Python/pystate.c:54]: Since this is an internal use, I think it would be OK to just write `runtime->gilstate.autoTLSkey._is_initialized = 0`.

However, I don't consider this a merge blocker either way - the `Py_tss_NEEDS_INIT` version also works, and we can make "Provide a nicer runtime TSS structpre-initialization API" a separate RFE.

File: Include/pythread.h
Comment (ncoghlan) [inline Include/pythread.h]: This ref can also be updated to 31370

File: Python/thread.c
Comment (ncoghlan) [inline Python/thread.c]: Suggested rephrasing for this comment: "Cross-platform components of TSS API implementation"

File: Python/thread_nt.h
Comment (ncoghlan) [inline Python/thread_nt.h]: May be worth mentioning PEP 539 specifically here.

File: Python/thread_nt.h
Comment (ncoghlan) [inline Python/thread_nt.h]: Rephrasing inline with the above suggestion: "Platform-specific components of TSS API implementation."

**Review by ncoghlan — APPROVED** [review]
Functionally, this is looking good to me now.

@ma8ma Would you be OK with my just going ahead and making the comment fixes I suggest? I'd also like to reword the What's New additions a bit, and add a new section to https://docs.python.org/3/c-api/init.html for Thread Specific Storage (since this is officially a public API now).

File: Python/pystate.c
Comment (ma8ma) [reply Python/pystate.c:54]: I'd like to maintain the status quo to indicate that `Py_tss_t` is an opaque type, this is the one of the few chance.  In fact, except for testing, there is only another example `Module/_tracemalloc.c` in CPython repo. This is also the reason that first I chose the assignment by struct, but not the feature of initializing struct field (I'm not sure it will be required...).

File: Python/thread_nt.h
Comment (ma8ma) [reply Python/thread_nt.h]: Sure, I fix comments in one go.

**Comment by ma8ma** [comment]
@ncoghlan Okay, I updated comments!

> add a new section to https://docs.python.org/3/c-api/init.html for Thread Specific Storage (...).

I'll try that too. (well, take time...)

> since this is officially a public API now

Does that mean the TSS API is part of Windows stable ABI? (i.e. commit 594ec29 again except for PyThread_tss_delete_value)

**Comment by ncoghlan** [comment]
@ma8ma Yes, the parts protected by `#ifdef Py_LIMITED_API+0 >= 0x03070000` would all qualify for inclusion in the Windows stable ABI.

However, even without that, things can be part of the public C API without being part of the stable ABI - it just means that they remain source-code compatible between releases, but not necessarily binary compatible. The `Py_tss_t` struct is an example of that here - it's not private (although the individual field names are), so if you're willing to recompile for every new CPython feature release, you can use it as a static type, rather than dynamically allocating it.

**Comment by ma8ma** [comment]
@ncoghlan Thank you for the explanation. I don't think there is any problem, so I add the TSS API in the Windows stable ABI.

**Comment by ma8ma** [comment]
I updated this PR to add API document. Sources of the API description are:

- [PEP 539](https://www.python.org/dev/peps/pep-0539/)
- comments on the code (5073d66)

Documents rendered by github:

- [c-api/init.rst](https://github.com/ma8ma/cpython/blob/pep539-tss-api/Doc/c-api/init.rst#thread-local-storage-support)
- [whatsnew/3.7.rst](https://github.com/ma8ma/cpython/blob/pep539-tss-api/Doc/whatsnew/3.7.rst#pep-539-a-new-c-api-for-thread-local-storage-in-cpython)

**Review by ncoghlan — CHANGES_REQUESTED** [review]
This is looking very good to me. I'm going to go through and make some edits to the docs updates, and add in the required limited API version check, but after that I think this should be ready to merge.

**Comment by ncoghlan** [comment]
Oops, the required macro version check is there, it just wasn't visible in the subset of changes I was looking at - so I'll just make the docs tweaks, and then this will be ready to merge.

**Comment by embray** [comment]
Great! I was just about to say, since `--without-threads` was removed it's impossible to compile on Cygwin without this now.  Is there anything else I can do to help or is this otherwise ready to go?

**Review by ncoghlan — APPROVED** [review]

**Comment by ncoghlan** [comment]
@embray I think it's good to go once the latest CI run finishes, but another set of eyes looking over the changes wouldn't hurt (if it's just docs comments though, then it would probably make sense to defer those to a new PR so we can restore Cygwin compatibility in the meantime)

**Comment by ncoghlan** [comment]
And merged - thank you @ma8ma for the PR!

**Comment by ma8ma** [comment]
@embray @ncoghlan Thank you for your advice too! :smile:

**Review by shaorongchen88 — COMMENTED** [review]
<deleted as the comment seemed like spam>

**Comment by pradyunsg** [comment]
> you will be put in the comfy chair!

This link doesn't seem to be working for me. I'm also watching other videos on YouTube currently so this seems to be this specific URL.

<details>
<summary>Screenshot -- click to view</summary>

screen shot 2018-07-20 at 8 37 51 am

</details>

**Review by qwhz — COMMENTED** [review]
- [ ] [](url)

**Pull request merged**
